File: 2023/src/bin/day5/main.rs
#![warn(clippy::pedantic)]
use std::{borrow::Cow, ops::Range};

use advent_2023::{BasicSolution, Solution};
use itertools::Itertools;
use winnow::{
    ascii::{dec_uint, line_ending},
    combinator::{not, opt, preceded, separated},
    seq, Parser,
};

struct Day;

#[derive(Debug, Clone)]
struct Mapping {
    source: Range<u64>,
    destination_start: u64,
}

#[derive(Debug, Clone)]
struct Almanac {
    seeds: Vec<u64>,
    all_mappings: Vec<Vec<Mapping>>,
}

#[derive(Debug, Clone)]
struct MappedRange {
    before: Option<Range<u64>>,
    overlap: Option<Range<u64>>,
    after: Option<Range<u64>>,
}

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample.txt");

    type Shared = Almanac;
    type Answer = u64;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 35;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 46;

    fn shared(input: &'static str) -> anyhow::Result<Self::Shared> {
        almanac.parse(input).map_err(anyhow::Error::msg)
    }

    fn part_a(almanac: Cow<Self::Shared>) -> anyhow::Result<Self::Answer> {
        almanac
            .seeds
            .iter()
            .map(|&seed| {
                almanac.all_mappings.iter().fold(seed, |acc, mappings| {
                    mappings
                        .iter()
                        .find_map(|mapping| {
                            mapping.source.contains(&acc).then(|| mapping.offset(acc))
                        })
                        .unwrap_or(acc)
                })
            })
            .min()
            .ok_or_else(|| anyhow::Error::msg("no seeds"))
    }

    fn part_b(almanac: Self::Shared) -> anyhow::Result<Self::Answer> {
        let seed_ranges = almanac.seeds.into_iter().tuples().map(|(a, b)| a..a + b);
        seed_ranges
            .flat_map(|seed_range| {
                almanac
                    .all_mappings
                    .iter()
                    .fold(vec![seed_range], |acc, mappings| {
                        acc.into_iter()
                            .flat_map(|input_range| multi_map_range(input_range, mappings))
                            .collect()
                    })
            })
            .map(|location_range| location_range.start)
            .min()
            .ok_or_else(|| anyhow::Error::msg("no location ranges"))
    }
}

fn multi_map_range(
    input_range: Range<u64>,
    mappings: &[Mapping],
) -> impl Iterator<Item = Range<u64>> {
    let mut overlaps = vec![];
    let mut leftovers = vec![input_range];
    let mut leftovers_queue = vec![];
    for mapping in mappings {
        for input_range in leftovers.drain(..) {
            let mapped = mapping.map_range(input_range);
            overlaps.extend(mapped.overlap);
            leftovers_queue.extend(mapped.before);
            leftovers_queue.extend(mapped.after);
        }
        std::mem::swap(&mut leftovers, &mut leftovers_queue);
    }
    overlaps.into_iter().chain(leftovers)
}

impl Mapping {
    fn offset(&self, value: u64) -> u64 {
        value + self.destination_start - self.source.start
    }

    fn map_range(&self, range: Range<u64>) -> MappedRange {
        let opt_range = |start, end| Some(start..end).filter(|r| !r.is_empty());
        let before = opt_range(range.start, self.source.start.min(range.end));
        let after = opt_range(self.source.end.max(range.start), range.end);
        let overlap = opt_range(
            range.start.max(self.source.start),
            range.end.min(self.source.end),
        )
        .map(|r| self.offset(r.start)..self.offset(r.end));
        MappedRange {
            before,
            overlap,
            after,
        }
    }
}

fn almanac(input: &mut &'static str) -> winnow::Result<Almanac> {
    seq! {Almanac {
        seeds: seeds,
        _: "\n\n",
        all_mappings: separated(7..=7, mappings, "\n\n"),
        _: opt("\n"),
    }}
    .parse_next(input)
}

fn seeds(input: &mut &str) -> winnow::Result<Vec<u64>> {
    preceded("seeds: ", separated(1.., u64, ' ')).parse_next(input)
}

fn mappings(input: &mut &str) -> winnow::Result<Vec<Mapping>> {
    preceded((not(line_ending), "\n"), separated(1.., mapping, "\n")).parse_next(input)
}

fn mapping(input: &mut &str) -> winnow::Result<Mapping> {
    seq! {Mapping {
        destination_start: u64,
        _: ' ',
        source: (u64, ' ', u64).map(|(start, _, len)| start..start + len),
    }}
    .parse_next(input)
}

fn u64(input: &mut &str) -> winnow::Result<u64> {
    dec_uint.parse_next(input)
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}

File: 2023/python/day5/main.py
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable, Generic, TypeVar

import toolz

SAMPLE_INPUT = """seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
"""
SAMPLE_INPUT_B = SAMPLE_INPUT
INPUT = Path("data.txt").read_text()


@dataclass
class Mapping:
    from_range: range
    dest: int


T = TypeVar("T")


@dataclass
class MappingResult(Generic[T]):
    mapped: T | None = None
    not_mapped: list[T] = field(default_factory=list)


def part_a(input: str):
    seeds, all_mappings = parse_input(input)
    locations = map_all(seeds, all_mappings, map_single)
    return min(locations)


def part_b(input: str):
    seeds, all_mappings = parse_input(input)
    seed_ranges = [range(a, a + b) for a, b in toolz.itertoolz.partition(2, seeds)]
    location_ranges = map_all(seed_ranges, all_mappings, map_range)
    return min(r.start for r in location_ranges)


def map_all(
    objects: list[T],
    all_mappings: list[list[Mapping]],
    mapping_function: Callable[[Mapping, T], MappingResult[T]],
) -> list[T]:
    for mappings in all_mappings:
        next_objects = []
        for mapping in mappings:
            leftovers = []
            for o in objects:
                mapping_result = mapping_function(mapping, o)
                if mapping_result.mapped is not None:
                    next_objects.append(mapping_result.mapped)
                for not_mapped in mapping_result.not_mapped:
                    leftovers.append(not_mapped)
            objects = leftovers
        objects = next_objects + objects
    return objects


def map_single(mapping: Mapping, input: int) -> MappingResult[int]:
    if input in mapping.from_range:
        return MappingResult(mapped=input + mapping.dest - mapping.from_range.start)
    return MappingResult(not_mapped=[input])


def map_range(mapping: Mapping, input: range) -> MappingResult[range]:
    before = range(input.start, min(input.stop, mapping.from_range.start))

    mapped = range(
        max(input.start, mapping.from_range.start)
        + mapping.dest
        - mapping.from_range.start,
        min(input.stop, mapping.from_range.stop)
        + mapping.dest
        - mapping.from_range.start,
    )

    after = range(max(input.start, mapping.from_range.stop), input.stop)

    return MappingResult(
        mapped=mapped if not_empty(mapped) else None,
        not_mapped=list(filter(not_empty, [before, after])),
    )


def not_empty(r: range) -> bool:
    return r.stop > r.start


def parse_input(input: str) -> tuple[list[int], list[list[Mapping]]]:
    seeds, *all_mappings = input.split("\n\n")
    seeds = [int(seed) for seed in seeds.split()[1:]]
    all_mappings = [
        [parse_mapping(mapping) for mapping in mappings.splitlines()[1:]]
        for mappings in all_mappings
    ]
    return seeds, all_mappings


def parse_mapping(line: str) -> Mapping:
    dest, start, length = line.split()
    dest = int(dest)
    start = int(start)
    length = int(length)
    return Mapping(from_range=range(start, start + length), dest=dest)


def test_part_a(capsys):
    assert part_a(SAMPLE_INPUT) == 35
    with capsys.disabled():
        print(f"Part A: {part_a(INPUT)}")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT_B) == 46
    with capsys.disabled():
        print(f"Part B: {part_b(INPUT)}")
